rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,315 @@
+# RFC 8288 - 2021-07-20 - CSV Enrichment
+
+Vector needs to allow users to enrich the events flowing through the topology
+using a CSV file. This RFC proposes adding a VRL function that can perform
+this lookup.
+
+## Scope
+
+The RFC will cover a simple VRL function that will lookup a single row from a
+CSV file using a set of conditions. This allows the user to map the data into
+the event using the full power of VRL.
+
+For MVP purposes, lookup will be a simple sequential scan. The first version
+of this function will not handle any indexing of this data. Above a certain size
+indexing will provide substantial performance benefits, so it is something that
+we should definitely consider at a future stage. Therefore, it is out of scope
+for this RFC.
+
+Encryption will not be covered. This is very much a requirement to be considered
+at a future stage, but is out of scope for this RFC.
+
+Likewise, reloading the enrichment data is likely to be needed, but will not be
+covered here. For the first implementation of this RFC, reloading the data will
+require Vector to be restarted.
+
+## Motivation
+
+Users would like to enrich events flowing through Vector with extra data
+provided by a CSV file.
+
+## Doc-level Proposal
+
+### Resources
```

Comment:
After thinking about @lukesteensen's proposal (join transform) and lightly discussing how the "resource" concept will evolve, I think a much better name for this is "table". It's just a table of data that we can build however we want (front loading, streaming, etc). To start, the CSV table will be all strings, but we can follow up with a schema enhancement if we desire. I think this naming is more intuitive since it aligns with database terminology and is a little more specific to it's purpose.

Reply:
Yes, that's a great idea. There was a vagueness in the name Resource that could cause people to think it's for all sorts of other things. Plus there is already the concept of Resource in the [source code](https://github.com/timberio/vector/blob/master/src/config/mod.rs#L466-L471) that is completely unrelated, so it makes it easier there too.